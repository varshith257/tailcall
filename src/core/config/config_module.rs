use std::collections::{HashMap, HashSet};
use std::ops::Deref;
use std::sync::Arc;

use derive_setters::Setters;
use jsonwebtoken::jwk::JwkSet;
use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};
use rustls_pki_types::{CertificateDer, PrivateKeyDer};

use crate::core::config::Config;
use crate::core::macros::MergeRight;
use crate::core::merge_right::MergeRight;
use crate::core::proto_reader::ProtoMetadata;
use crate::core::rest::{EndpointSet, Unchecked};

/// A wrapper on top of Config that contains all the resolved extensions and
/// computed values.
#[derive(Clone, Debug, Default, Setters, MergeRight)]
pub struct ConfigModule {
    pub config: Config,
    pub extensions: Extensions,
    pub input_types: HashSet<String>,
    pub output_types: HashSet<String>,
    pub interface_types: HashSet<String>,
}

#[derive(Clone, Debug, Default)]
pub struct Content<A> {
    pub id: Option<String>,
    pub content: A,
}

impl<A> Deref for Content<A> {
    type Target = A;
    fn deref(&self) -> &Self::Target {
        &self.content
    }
}

/// Extensions are meta-information required before we can generate the
/// blueprint. Typically, this information cannot be inferred without performing
/// an IO operation, i.e., reading a file, making an HTTP call, etc.
#[derive(Clone, Debug, Default, MergeRight)]
pub struct Extensions {
    /// Contains the file descriptor set resolved from the links to proto files
    pub grpc_file_descriptors: HashMap<String, FileDescriptorProto>,

    /// Contains the contents of the JS file
    pub script: Option<String>,

    /// Contains the certificate used on HTTP2 with TLS
    pub cert: Vec<CertificateDer<'static>>,

    /// Contains the key used on HTTP2 with TLS
    pub keys: Arc<Vec<PrivateKeyDer<'static>>>,

    /// Contains the endpoints
    pub endpoint_set: EndpointSet<Unchecked>,

    pub htpasswd: Vec<Content<String>>,

    pub jwks: Vec<Content<JwkSet>>,
}

impl Extensions {
    pub fn add_proto(&mut self, metadata: ProtoMetadata) {
        for file in metadata.descriptor_set.file {
            self.grpc_file_descriptors
                .insert(file.name().to_string(), file);
        }
    }

    pub fn get_file_descriptor_set(&self) -> FileDescriptorSet {
        FileDescriptorSet { file: self.grpc_file_descriptors.values().cloned().collect() }
    }

    pub fn has_auth(&self) -> bool {
        !self.htpasswd.is_empty() || !self.jwks.is_empty()
    }
}

impl MergeRight for FileDescriptorSet {
    fn merge_right(mut self, other: Self) -> Self {
        self.file.extend(other.file);

        self
    }
}

impl Deref for ConfigModule {
    type Target = Config;
    fn deref(&self) -> &Self::Target {
        &self.config
    }
}

pub struct Resolution {
    pub input: String,
    pub output: String,
}

fn insert_resolution(
    mut map: HashMap<String, Resolution>,
    current_name: &str,
    resolution: Resolution,
) -> HashMap<String, Resolution> {
    if resolution.input.eq(&resolution.output) {
        tracing::warn!(
            "Unable to resolve input and output type: {}",
            resolution.input
        );
    }

    if !map.contains_key(current_name) {
        map.entry(current_name.to_string()).or_insert(resolution);
    }

    map
}

impl ConfigModule {
    /// This function resolves the ambiguous types by renaming the input and
    /// output types. The resolver function should return a Resolution
    /// object containing the new input and output types.
    /// The function will return a new ConfigModule with the resolved types.
    pub fn resolve_ambiguous_types(mut self, resolver: impl Fn(&str) -> Resolution) -> Self {
        let mut resolution_map = HashMap::new();

        // iterate over intersection of input and output types
        for current_name in self.input_types.intersection(&self.output_types) {
            let resolution = resolver(current_name);

            resolution_map = insert_resolution(resolution_map, current_name, resolution);

            if let Some(ty) = self.config.types.get(current_name) {
                for field in ty.fields.values() {
                    for args in field.args.values() {
                        // if arg is of output type then it should be changed to that of newly
                        // created input type.
                        if self.output_types.contains(&args.type_of)
                            && !resolution_map.contains_key(&args.type_of)
                        {
                            let resolution = resolver(args.type_of.as_str());
                            resolution_map = insert_resolution(
                                resolution_map,
                                args.type_of.as_str(),
                                resolution,
                            );
                        }
                    }
                }
            }
        }

        // insert newly created types to the config.
        for (current_name, resolution) in &resolution_map {
            let input_name = &resolution.input;
            let output_name = &resolution.output;

            let og_ty = self.config.types.get(current_name).cloned();

            // remove old types
            self.config.types.remove(current_name);
            self.input_types.remove(current_name);
            self.output_types.remove(current_name);

            // add new types
            if let Some(og_ty) = og_ty {
                self.config.types.insert(input_name.clone(), og_ty.clone());
                self.input_types.insert(input_name.clone());

                self.config.types.insert(output_name.clone(), og_ty);
                self.output_types.insert(output_name.clone());
            }
        }

        let keys = self.config.types.keys().cloned().collect::<Vec<String>>();

        for k in keys {
            if let Some(ty) = self.config.types.get_mut(&k) {
                for field in ty.fields.values_mut() {
                    if let Some(resolution) = resolution_map.get(&field.type_of) {
                        if self.output_types.contains(&k) {
                            field.type_of.clone_from(&resolution.output);
                        } else if self.input_types.contains(&k) {
                            field.type_of.clone_from(&resolution.input);
                        }
                    }
                    for arg in field.args.values_mut() {
                        if let Some(resolution) = resolution_map.get(&arg.type_of) {
                            arg.type_of.clone_from(&resolution.input);
                        }
                    }
                }
            }
        }

        self
    }
}

impl From<Config> for ConfigModule {
    fn from(config: Config) -> Self {
        let input_types = config.input_types();
        let output_types = config.output_types();
        let interface_types = config.interface_types();

        ConfigModule {
            config,
            input_types,
            output_types,
            interface_types,
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use maplit::hashset;
    use pretty_assertions::assert_eq;

    use crate::core::config::{Config, ConfigModule, Resolution, Type};
    use crate::core::generator::Source;

    fn build_qry(mut config: Config) -> Config {
        let mut query = Type::default();
        let mut field1 =
            crate::core::config::Field { type_of: "Type1".to_string(), ..Default::default() };

        let arg1 = crate::core::config::Arg { type_of: "Type1".to_string(), ..Default::default() };

        field1.args.insert("arg1".to_string(), arg1);

        let arg2 = crate::core::config::Arg { type_of: "Type2".to_string(), ..Default::default() };

        field1.args.insert("arg2".to_string(), arg2);

        let mut field2 = field1.clone();
        field2.type_of = "Type2".to_string();

        query.fields.insert("field1".to_string(), field1);
        query.fields.insert("field2".to_string(), field2);

        config.types.insert("Query".to_string(), query);
        config = config.query("Query");

        config
    }

    #[test]
    fn test_resolve_ambiguous_types() {
        // Create a ConfigModule instance with ambiguous types
        let mut config = Config::default();

        let mut type1 = Type::default();
        let mut type2 = Type::default();
        let mut type3 = Type::default();

        type1.fields.insert(
            "name".to_string(),
            crate::core::config::Field::default().type_of("String".to_string()),
        );

        type2.fields.insert(
            "ty1".to_string(),
            crate::core::config::Field::default().type_of("Type1".to_string()),
        );

        type3.fields.insert(
            "ty1".to_string(),
            crate::core::config::Field::default().type_of("Type1".to_string()),
        );
        type3.fields.insert(
            "ty2".to_string(),
            crate::core::config::Field::default().type_of("Type2".to_string()),
        );

        config.types.insert("Type1".to_string(), type1);
        config.types.insert("Type2".to_string(), type2);
        config.types.insert("Type3".to_string(), type3);

        config = build_qry(config);

        let mut config_module = ConfigModule::from(config);

        let resolver = |type_name: &str| -> Resolution {
            Resolution {
                input: format!("{}Input", type_name),
                output: format!("{}Output", type_name),
            }
        };

        config_module = config_module.resolve_ambiguous_types(resolver);

        let actual = config_module
            .config
            .types
            .keys()
            .map(|s| s.as_str())
            .collect::<HashSet<_>>();

        let expected = hashset![
            "Query",
            "Type1Input",
            "Type1Output",
            "Type2Input",
            "Type2Output",
            "Type3"
        ];

        assert_eq!(actual, expected);
    }
    #[tokio::test]
    async fn test_resolve_ambiguous_news_types() -> anyhow::Result<()> {
        let gen = crate::core::generator::Generator::init(crate::core::runtime::test::init(None));
        let news = tailcall_fixtures::protobuf::NEWS;
        let config_module = gen.read_all(Source::Proto, &[news], "Query").await?;
        let actual = config_module
            .config
            .types
            .keys()
            .map(|s| s.as_str())
            .collect::<HashSet<_>>();

        let expected = hashset![
            "Query",
            "news__News",
            "news__NewsList",
            "news__NewsInput",
            "news__NewsId",
            "news__MultipleNewsId"
        ];
        assert_eq!(actual, expected);
        Ok(())
    }
}
