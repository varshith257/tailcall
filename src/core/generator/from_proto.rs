use std::collections::BTreeSet;

use anyhow::{bail, Result};
use derive_setters::Setters;
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto,
};

use super::graphql_type::{GraphQLType, Unparsed};
use crate::core::config::{Arg, Config, Enum, Field, Grpc, Tag, Type};

/// Assists in the mapping and retrieval of proto type names to custom formatted
/// strings based on the descriptor type.
#[derive(Setters)]
struct Context {
    /// The current proto package name.
    namespace: Vec<String>,

    /// Final configuration that's being built up.
    config: Config,

    /// Root GraphQL query type
    query: String,
}

impl Context {
    fn new(query: &str) -> Self {
        Self {
            query: query.to_string(),
            namespace: Default::default(),
            config: Default::default(),
        }
    }

    /// Resolves the actual name and inserts the type.
    fn insert_type(mut self, name: String, ty: Type) -> Self {
        self.config.types.insert(name.to_string(), ty);
        self
    }

    /// Processes proto enum types.
    fn append_enums(mut self, enums: &Vec<EnumDescriptorProto>) -> Self {
        for enum_ in enums {
            let enum_name = enum_.name();

            let variants = enum_
                .value
                .iter()
                .map(|v| GraphQLType::new(v.name()).into_enum_variant().to_string())
                .collect::<BTreeSet<String>>();

            let type_name = GraphQLType::new(enum_name)
                .extend(self.namespace.as_slice())
                .into_enum()
                .to_string();
            self.config
                .enums
                .insert(type_name, Enum { variants, doc: None });
        }
        self
    }

    /// Processes proto message types.
    fn append_msg_type(mut self, messages: &Vec<DescriptorProto>) -> Result<Self> {
        for message in messages {
            let msg_name = message.name();
            if let Some(options) = message.options.as_ref() {
                if options.map_entry.unwrap_or_default() {
                    continue;
                }
            }

            // first append the name of current message as namespace
            self.namespace.push(msg_name.to_string());
            self = self.append_enums(&message.enum_type);
            self = self.append_msg_type(&message.nested_type)?;
            // then drop it after handling nested types
            self.namespace.pop();

            let msg_type = GraphQLType::new(msg_name)
                .extend(self.namespace.as_slice())
                .into_object_type();

            let mut ty = Type::default();
            for field in message.field.iter() {
                let field_name = GraphQLType::new(field.name())
                    .extend(self.namespace.as_slice())
                    .into_field();

                let mut cfg_field = Field::default();

                let label = field.label().as_str_name().to_lowercase();
                cfg_field.list = label.contains("repeated");
                cfg_field.required = label.contains("required") || cfg_field.list;

                if field.type_name.is_some() {
                    // for non-primitive types
                    let type_of = graphql_type_from_ref(field.type_name())?
                        .into_object_type()
                        .to_string();

                    cfg_field.type_of = type_of;
                } else {
                    let type_of = convert_primitive_type(field.r#type().as_str_name());
                    if type_of.eq("JSON") {
                        cfg_field.list = false;
                        cfg_field.required = false;
                    }
                    cfg_field.type_of = type_of;
                }

                ty.fields.insert(field_name.to_string(), cfg_field);
            }

            ty.tag = Some(Tag { id: msg_type.id() });

            self = self.insert_type(msg_type.to_string(), ty);
        }
        Ok(self)
    }

    /// Processes proto service definitions and their methods.
    fn append_query_service(mut self, services: &Vec<ServiceDescriptorProto>) -> Result<Self> {
        if services.is_empty() {
            return Ok(self);
        }

        for service in services {
            let service_name = service.name();
            for method in &service.method {
                let field_name = GraphQLType::new(method.name())
                    .extend(self.namespace.as_slice())
                    .push(service_name)
                    .into_method();

                let mut cfg_field = Field::default();
                let mut body = None;

                if let Some(graphql_type) = get_input_type(method.input_type())? {
                    let key = graphql_type.clone().into_field().to_string();
                    let type_of = graphql_type.into_object_type().to_string();
                    let val = Arg {
                        type_of,
                        list: false,
                        required: true,
                        /* Setting it not null by default. There's no way to infer this
                         * from proto file */
                        doc: None,
                        modify: None,
                        default_value: None,
                    };

                    body = Some(format!("{{{{.args.{key}}}}}"));
                    cfg_field.args.insert(key, val);
                }

                let output_ty = get_output_type(method.output_type())?
                    .into_object_type()
                    .to_string();
                cfg_field.type_of = output_ty;
                cfg_field.required = true;

                cfg_field.grpc = Some(Grpc {
                    base_url: None,
                    body,
                    group_by: vec![],
                    headers: vec![],
                    method: field_name.id(),
                });

                let ty = self
                    .config
                    .types
                    .entry(self.query.clone())
                    .or_insert_with(|| {
                        self.config.schema.query = Some(self.query.clone());
                        Type::default()
                    });

                ty.fields.insert(field_name.to_string(), cfg_field);
            }
        }
        Ok(self)
    }
}

fn graphql_type_from_ref(name: &str) -> Result<GraphQLType<Unparsed>> {
    if !name.starts_with('.') {
        bail!("Expected fully-qualified name for reference type but got {name}. This is a bug!");
    }

    let name = &name[1..];

    if let Some((package, name)) = name.rsplit_once('.') {
        Ok(GraphQLType::new(name).push(package))
    } else {
        Ok(GraphQLType::new(name))
    }
}

/// Converts proto field types to a custom format.
fn convert_primitive_type(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "int64" | "fixed32" | "fixed64" | "uint32" | "uint64" => "Int",
        "bool" => "Boolean",
        "string" | "bytes" => "String",
        "message" => "JSON", // JSON scalar is preloaded by tailcall, so there is no need to
        // explicitly define it in the config.
        x => x,
    }
    .to_string()
}

/// Determines the output type for a service method.
fn get_output_type(output_ty: &str) -> Result<GraphQLType<Unparsed>> {
    // type, required
    match output_ty {
        ".google.protobuf.Empty" => {
            // If it's no response is expected, we return an Empty scalar type
            Ok(GraphQLType::new("Empty"))
        }
        _ => {
            // Setting it not null by default. There's no way to infer this from proto file
            graphql_type_from_ref(output_ty)
        }
    }
}

fn get_input_type(input_ty: &str) -> Result<Option<GraphQLType<Unparsed>>> {
    match input_ty {
        ".google.protobuf.Empty" | "" => Ok(None),
        _ => graphql_type_from_ref(input_ty).map(Some),
    }
}

/// The main entry point that builds a Config object from proto descriptor sets.
pub fn from_proto(descriptor_sets: &[FileDescriptorSet], query: &str) -> Result<Config> {
    let mut ctx = Context::new(query);
    for descriptor_set in descriptor_sets.iter() {
        for file_descriptor in descriptor_set.file.iter() {
            ctx.namespace = vec![file_descriptor.package().to_string()];

            ctx = ctx
                .append_enums(&file_descriptor.enum_type)
                .append_msg_type(&file_descriptor.message_type)?
                .append_query_service(&file_descriptor.service)?;
        }
    }

    let unused_types = ctx.config.unused_types();
    ctx.config = ctx.config.remove_types(unused_types);

    Ok(ctx.config)
}

#[cfg(test)]
mod test {
    use anyhow::Result;
    use prost_reflect::prost_types::FileDescriptorSet;
    use tailcall_fixtures::protobuf;

    use super::*;
    use crate::core::config::{ConfigModule, Resolution};

    fn compile_protobuf(files: &[&str]) -> Result<FileDescriptorSet> {
        Ok(protox::compile(files, [protobuf::SELF])?)
    }

    #[test]
    fn test_from_proto() -> Result<()> {
        // news_enum.proto covers:
        // test for mutation
        // test for empty objects
        // test for optional type
        // test for enum
        // test for repeated fields
        // test for a type used as both input and output
        // test for two types having same name in different packages

        let set =
            compile_protobuf(&[protobuf::NEWS, protobuf::GREETINGS_A, protobuf::GREETINGS_B])?;
        let result = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(result);

        Ok(())
    }

    #[test]
    fn test_from_proto_no_pkg_file() -> Result<()> {
        let set = compile_protobuf(&[protobuf::NEWS_NO_PKG])?;
        let result = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(result);
        Ok(())
    }

    #[test]
    fn test_from_proto_no_service_file() -> Result<()> {
        let set = compile_protobuf(&[protobuf::NEWS_NO_SERVICE])?;
        let result = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(result);

        Ok(())
    }

    #[test]
    fn test_greetings_proto_file() -> Result<()> {
        let set = compile_protobuf(&[protobuf::GREETINGS, protobuf::GREETINGS_MESSAGE]).unwrap();
        let result = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(result);

        Ok(())
    }

    #[test]
    fn test_config_from_sdl() -> Result<()> {
        let set =
            compile_protobuf(&[protobuf::NEWS, protobuf::GREETINGS_A, protobuf::GREETINGS_B])?;

        let set1 = compile_protobuf(&[protobuf::NEWS])?;
        let set2 = compile_protobuf(&[protobuf::GREETINGS_A])?;
        let set3 = compile_protobuf(&[protobuf::GREETINGS_B])?;

        let actual = from_proto(&[set.clone()], "Query")?.to_sdl();
        let expected = from_proto(&[set1, set2, set3], "Query")?.to_sdl();

        pretty_assertions::assert_eq!(actual, expected);
        Ok(())
    }

    #[test]
    fn test_required_types() -> Result<()> {
        // required fields are deprecated in proto3 (https://protobuf.dev/programming-guides/dos-donts/#add-required)
        // this example uses proto2 to test the same.
        // for proto3 it's guaranteed to have a default value (https://protobuf.dev/programming-guides/proto3/#default)
        // and our implementation (https://github.com/tailcallhq/tailcall/pull/1537) supports default values by default.
        // so we do not need to explicitly mark fields as required.

        let set = compile_protobuf(&[protobuf::PERSON])?;
        let config = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(config);
        Ok(())
    }

    #[test]
    fn test_movies() -> Result<()> {
        let set = compile_protobuf(&[protobuf::MOVIES])?;
        let config = from_proto(&[set], "Query")?;
        let config_module = ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
            input: format!("{}Input", v),
            output: v.to_owned(),
        });
        let config = config_module.to_sdl();
        insta::assert_snapshot!(config);

        Ok(())
    }

    #[test]
    fn test_nested_types() -> Result<()> {
        let set = compile_protobuf(&[protobuf::NESTED_TYPES])?;
        let config = from_proto(&[set], "Query")?.to_sdl();
        insta::assert_snapshot!(config);
        Ok(())
    }
}
