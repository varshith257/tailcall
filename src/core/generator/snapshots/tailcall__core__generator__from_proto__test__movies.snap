---
source: src/core/generator/from_proto.rs
expression: config
---
schema @server @upstream {
  query: Query
}

input google__protobuf__DurationInput @tag(id: "google.protobuf.Duration") {
  nanos: Int
  seconds: Int
}

input google__protobuf__Int32ValueInput @tag(id: "google.protobuf.Int32Value") {
  value: Int
}

input google__protobuf__StringValue @tag(id: "google.protobuf.StringValue") {
  value: String
}

input google__protobuf__TimestampInput @tag(id: "google.protobuf.Timestamp") {
  nanos: Int
  seconds: Int
}

input movies__MovieInput @tag(id: "movies.Movie") {
  cast: [String]!
  duration: google__protobuf__DurationInput
  genre: movies__GenreInput
  name: String
  rating: Float
  time: google__protobuf__TimestampInput
  year: google__protobuf__Int32ValueInput
}

input movies__MovieRequest @tag(id: "movies.MovieRequest") {
  movie: movies__MovieInput
}

input movies__SearchByCastRequest @tag(id: "movies.SearchByCastRequest") {
  castName: google__protobuf__StringValue
}

enum movies__Genre {
  ACTION
  DRAMA
  UNSPECIFIED
}

type Query {
  movies__AnotherExample__GetMovies(movieRequest: movies__MovieRequest!): movies__MoviesResult! @grpc(body: "{{.args.movieRequest}}", method: "movies.AnotherExample.GetMovies")
  movies__AnotherExample__SearchMoviesByCast(searchByCastRequest: movies__SearchByCastRequest!): movies__Movie! @grpc(body: "{{.args.searchByCastRequest}}", method: "movies.AnotherExample.SearchMoviesByCast")
  movies__Example__GetMovies(movieRequest: movies__MovieRequest!): movies__MoviesResult! @grpc(body: "{{.args.movieRequest}}", method: "movies.Example.GetMovies")
  movies__Example__SearchMoviesByCast(searchByCastRequest: movies__SearchByCastRequest!): movies__Movie! @grpc(body: "{{.args.searchByCastRequest}}", method: "movies.Example.SearchMoviesByCast")
}

type google__protobuf__Duration @tag(id: "google.protobuf.Duration") {
  nanos: Int
  seconds: Int
}

type google__protobuf__Int32Value @tag(id: "google.protobuf.Int32Value") {
  value: Int
}

type google__protobuf__Timestamp @tag(id: "google.protobuf.Timestamp") {
  nanos: Int
  seconds: Int
}

type movies__Movie @tag(id: "movies.Movie") {
  cast: [String]!
  duration: google__protobuf__Duration
  genre: movies__Genre
  name: String
  rating: Float
  time: google__protobuf__Timestamp
  year: google__protobuf__Int32Value
}

type movies__MoviesResult @tag(id: "movies.MoviesResult") {
  result: [movies__Movie]!
}
